### PR TITLE
Feature/Build spatial index based on google::s2

### PR DESCRIPTION
**Pull request by CodeBear801**
# Issue
https://github.com/Telenav/osrm-backend/issues/239

- implement `indexer` to build spatial index based on `points`, implement a specified version of `indexer` based on google::s2
- could answer `nearbyQuery` for any given point
- Most logic are inside `s2indexer`, it could 
   + construct `spatial index` with `Build()`, loads data from json, and iterate each points, find a group of cell id contains such point, in [minlevel, maxlevel] range
   + implement `Dump` and `Load` function, could record pre-processing data to files, so next time just need load data back to re-build spatial information
   + For `queryNearByPoints`, it will first find all cells covers query region, in [minlevel, maxlevel] range, then retrieves points data from those cells, you could go to here for more information: https://github.com/Telenav/osrm-backend/issues/236#issuecomment-603533484
    

## Tasklist

 - [x] implement code
 - [x] add tests 
 - [ ] review

### CODE REVIEW

File: integration/cmd/chargestations-connectivity-gen/doc.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// => iterate each point
+// => generate a circle(s2::cap), find all cellids intersect with that circle
+// => retrieve all ids
+// => generate result of id(from), ids(all ids in certain distance)
+// stage 3:
+// => load data from file
+// => for each line, its formid and all other ids
+// => calculate distance between fromid and all other ids
+// => sort result based on distance and write back to file
+package main
```
Comment (CodeBear801) [inline integration/cmd/chargestations-connectivity-gen/doc.go]: Just put comments here, to be implemented.

**Comment by CodeBear801** [comment]
Build failed with following information, didn't notice it on local, try to reproduce with docker.
```
##[error]Process completed with exit code 2.
```

**Comment by wangyoucao577** [comment]
> Build failed with following information, didn't notice it on local, try to reproduce with docker.
> 
> ```
> ##[error]Process completed with exit code 2.
> ```
> 
> update: I didn't met such error inside docker
> 
> ```
> go build -v ./...
> ```
> 
> image
> 
> May be not this command?

Here's the problem: 
```
?   	github.com/Telenav/osrm-backend/integration/wayidsmap	[no test files]
# github.com/Telenav/osrm-backend/integration/cmd/chargestations-connectivity-gen
runtime.main_main·f: function main is undeclared in the main package
The command '/bin/sh -c cd /workspace/osrm-backend/integration &&   go test -short -v ./... &&   go install ./... &&   ls -lh /workspace/go/bin' returned a non-zero code: 2
```
More in link: https://github.com/Telenav/osrm-backend/pull/248/checks?check_run_id=532257078

**Comment by CodeBear801** [comment]
> > Build failed with following information, didn't notice it on local, try to reproduce with docker.
> > ```
> > ##[error]Process completed with exit code 2.
> > ```
> > 
> > 
> > update: I didn't met such error inside docker
> > ```
> > go build -v ./...
> > ```
> > 
> > 
> > image
> > May be not this command?
> 
> Here's the problem:
> 
> ```
> ?   	github.com/Telenav/osrm-backend/integration/wayidsmap	[no test files]
> # github.com/Telenav/osrm-backend/integration/cmd/chargestations-connectivity-gen
> runtime.main_main·f: function main is undeclared in the main package
> The command '/bin/sh -c cd /workspace/osrm-backend/integration &&   go test -short -v ./... &&   go install ./... &&   ls -lh /workspace/go/bin' returned a non-zero code: 2
> ```
> 
> More in link: https://github.com/Telenav/osrm-backend/pull/248/checks?check_run_id=532257078

Thanks, found that immediately after dinner, hard to find this due to lake of energy :)

File: integration/cmd/chargestations-connectivity-gen/main.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package main
+
+import (
+	"fmt"
+)
+
+func main() {
+	fmt.Print("Hello World!")
```
Comment (wangyoucao577) [inline integration/cmd/chargestations-connectivity-gen/main.go]: Nit: `//TODO: `

File: integration/service/poiloader/poi_format.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
+package poiloader
+
+type Element struct {
```
Comment (wangyoucao577) [inline integration/service/poiloader/poi_format.go]: Comment

File: integration/service/poiloader/poi_format.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	CategoryIDGather  *CategoryIDGather `json:"category_id_gather"`
+	ChainGather       string            `json:"chain_gather"`
+	RawCategoryGather string            `json:"raw_category_gather"`
+	ChildGather       string            `json:"child_gather"`
+	ParentGather      string            `json:"parent_gather"`
+	Hilbert           float64           `json:"hilbert"`
+	Amenity           Amenity           `json:"amenity"`
+}
+
+type Amenity string
```
Comment (wangyoucao577) [inline integration/service/poiloader/poi_format.go]: Comment. Be aware of any warning in your IDE, :)

File: integration/service/poiloader/loader.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package poiloader
```
Comment (wangyoucao577) [inline integration/service/poiloader/loader.go]: It's more like a functional package for service `oasis`, right? Not a new service.

File: integration/service/spatialindexer/iterface.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package spatialindexer
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/iterface.go]: `interface.go`?

File: integration/service/spatialindexer/iterface.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// PointID defines ID for given point(location, point of interest)
+// Only the data used for pre-processing contains valid PointID
+type PointID int64
+
+// Finder answers special query
+type Finder interface {
+
+	// FindNearByIDs returns a group of points near to given center location
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/iterface.go]: comment not match

File: integration/service/spatialindexer/s2indexer/indexer.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/golang/glog"
+)
+
+type s2Indexer struct {
+	cellID2PointIDs  map[s2.CellID][]spatialindexer.PointID
+	pointID2Location map[spatialindexer.PointID]spatialindexer.Location
+}
+
+// NewS2Indexer generates spatial indexer based on google s2
+func NewS2Indexer() *s2Indexer {
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/s2indexer/indexer.go]: `New()`. Also, the `s2Indexer` isn't exported, can it be used by caller?

Comment (CodeBear801) [reply integration/service/spatialindexer/s2indexer/indexer.go]: So far, I don't expect it be used by caller.  Caller's code will looks like

```go
indexer := NewS2Indexer.Build().Dump(xxx)
```
I will try to put `indexer` related algorithm in side this package.   Will try.

File: integration/service/poiloader/poi_format.go (hunk 4 above)
Comment (CodeBear801) [reply integration/service/poiloader/poi_format.go]: Copy that.  Yes.
Also be aware of automatically generated code :)

File: integration/service/poiloader/loader.go (hunk 5 above)
Comment (CodeBear801) [reply integration/service/poiloader/loader.go]: Originally, I thought this package might be used by other package.  But you are right, currently the only user is inside of spatialIndexer.  Will put it there not under /service

File: integration/service/spatialindexer/iterface.go (hunk 6 above)
Comment (CodeBear801) [reply integration/service/spatialindexer/iterface.go]: interface.go

File: integration/service/spatialindexer/s2indexer/indexer.go (hunk 8 above)
Comment (CodeBear801) [reply integration/service/spatialindexer/s2indexer/indexer.go]: <del>
> So far, I don't expect it be used by caller. Caller's code will looks like
> 
> ```go
> indexer := NewS2Indexer.Build().Dump(xxx)
> ```
> 
> I will try to put `indexer` related algorithm in side this package.
</del>

**Comment by CodeBear801** [comment]
Will change `poi` and `Points` to `Place` for later commit.

`Place` here identify a bunch of points(place, poi) be pre-processed, build spatial index and connectivity.  Its different with random `point` on Earth.

File: integration/service/spatialindexer/doc.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
+// package spatialindexer answers query of nearest points(place, point of interest) for conditions
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/doc.go]: `package`->`Package` in comments, start with capitalized letter. 

See more details in https://github.com/golang/go/wiki/CodeReviewComments#package-comments

File: integration/service/spatialindexer/poiloader/poi_format.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+const (
+	ChargingStation Amenity = "charging_station"
+)
+
+// Country records IOS code
+type Country string
+
+const (
+	Usa Country = "USA"
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/poiloader/poi_format.go]: This also needs comments. 
Every capitalized stuff should have a comment.

File: integration/service/spatialindexer/doc.go (hunk 9 above)
Comment (CodeBear801) [reply integration/service/spatialindexer/doc.go]: Copy that.

**Pull request merged**